torch/csrc/jit/script/compiler.cpp

```diff
@@ -877,34 +875,57 @@ std::shared_ptr<SugaredValue> SimpleValue::attr(SourceRange loc, Method & m, con
 }
 
 
-void defineMethodsInModule(Module & m, const std::vector<Def>& definitions, const Resolver& resolver, SugaredValuePtr self) {
+void defineMethodsInModule(Module & m, const std::vector<Def>& definitions, const std::vector<Resolver>& resolvers, SugaredValuePtr self) {
   FunctionTable table;
-  for(auto def : definitions) {
+  JIT_ASSERT(definitions.size() == resolvers.size());
+  auto resolver_it = resolvers.begin();
+  std::vector<Method*> methods;
+  for(Def def : definitions) {
     const std::string& name = def.name().name();
-    Method& method = m.create_method(name);
-    to_ir(def, table, resolver, self,  method);
-    auto result = table.emplace(name, method);
-    if(!result.second) {
-      throw ErrorReport(def) << "duplicate definition of function '" << name << "'";
+    Resolver resolver = *resolver_it++;
+    auto creator = [def, &table, resolver, self](Method& method) {
+      to_ir(def, table, resolver, self,  method);
+    };
+    Method& method = m.create_method(name, creator);
+    // if self is defined, then these are methods and do not go into the global namespace
+    // otherwise, they get defined together so we add them to the function table
+    // so the methods can see each other
```

Comment:
I'm not sure if I understand this comment. How do the functions see each other if you don't put them in the table (the method case)?

Reply:
Hmm so this function has a silent invariant that `self` is a `ModuleValue` corresponding to `m`? Maybe we should remove it as an argument and instantiate the sugared value for it ourselves?